### PR TITLE
Nima WebClient configuration for timeouts and keepAlive

### LINKED ISSUE #6961: Nima WebClient Shortcuts for timeouts
In 3.x we had this: https://helidon.io/docs/v3/apidocs/io.helidon.webclient/io/helidon/webclient/WebClient.Builder.html#connectTimeout(long,java.util.concurrent.TimeUnit)

```java
WebClient client = WebClient.builder()
                            .connectTimeout(30, TimeUnit.SECONDS)
                            .readTimeout(10, TimeUnit.SECONDS)
                            .build();

WebClientResponse response = client.get()
                                   .connectTimeout(30, TimeUnit.SECONDS)
                                   .readTimeout(10, TimeUnit.SECONDS)
                                   .submit()
                                   .await();
```

### LINKED ISSUE #6979: Nima WebClient Configuration for keepAlive

### PR DESCRIPTION
**Pull request by aserkes**
Fixes #6961 
Fixes #6979

### CODE REVIEW

**Review by romain-grecourt — COMMENTED** [review]
Here is what can be done with the reactive WebClient in 3.x and 2.x :

```java
WebClient client = WebClient.builder()
                            .connectTimeout(30, TimeUnit.SECONDS)
                            .readTimeout(10, TimeUnit.SECONDS)
                            .build();

WebClientResponse response = client.get()
                                   .connectTimeout(30, TimeUnit.SECONDS)
                                   .readTimeout(10, TimeUnit.SECONDS)
                                   .submit()
                                   .await();
```

Here is some Nima based code:
```
Http1Client client =
        Http1Client.builder()
                   .channelOptions(SocketOptions
                           .builder()
                           .connectTimeout(Duration.ofSeconds(30))
                           .readTimeout(Duration.ofSeconds(10))
                           .build())
                   //.connectTimeout(Duration.ofSeconds(30)) <-- shortcut
                   //.readTimeout(Duration.ofSeconds(10))    <-- shortcut
                   .build();
Http1ClientResponse response = client.get()
                                     //.connectTimeout(Duration.ofSeconds(30))  <-- if possible
                                     //.readTimeout(Duration.ofSeconds(10))     <-- if possible
                                     .request();
```

The commented out code above is what needs to be implemented.

**Comment by aserkes** [comment]
If I understand it correctly, these timeouts related to the timeouts from `SocketOptions`. Therefore, to add the shortcuts and make these timeouts consistent with the timeouts from `SocketOptions`, as an option, I need to change `SocketOptions` and add a getter for `socketOptions` field in this class. Is it possible to change `SocketOptions`?

**Comment by romain-grecourt** [comment]
Here are some rough steps

- Add a new field to `WebClient.Builder`: `SocketOptions.Builder channelOptionsBuilder`.
- In `WebClient.Builder.build()` if `channelOptions` is null, then assign a value using  `channelOptionsBuilder.build()`
- Update `LoomClient` constructor to remove the null check since channelOptionsBuilder is now never null
- Add the following methods to `WebBuilder.Builder`:
```java
public B readTimeout(Duration readTimeout) {
    this.channelOptionsBuilder.readTimeout(readTimeout);
    return (B) this;
}

public B connectTimeout(Duration connectTimeout) {
    this.channelOptionsBuilder.connectTimeout(connectTimeout);
    return (B) this;
}
```

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 1]
```diff
@@ -134,7 +139,11 @@ public B tls(Supplier<Tls> tls) {
          * @return updated builder
          */
         public B channelOptions(SocketOptions channelOptions) {
-            this.channelOptions = channelOptions;
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:137]: I don't think we want to modify this method.
If an instance of `SocketOptions` is provided, we use it as-is and that means we discard the `SocketOptions.Builder`.

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 2]
```diff
@@ -67,12 +70,14 @@ static <T> T builder(Protocol<T> protocol) {
...
         private URI baseUri;
         private Tls tls;
-        private SocketOptions channelOptions;
+        private final SocketOptions.Builder socketOptionsBuilder = SocketOptions.builder();
         private DnsResolver dnsResolver;
         private DnsAddressLookup dnsAddressLookup;
         private boolean followRedirect;
         private int maxRedirect;
         private WritableHeaders<?> defaultHeaders = WritableHeaders.create();
+        private Duration connectTimeout;
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:79]: We should not need to store these as fields, we can simply update the `SocketOptions.Builder`

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 3]
```diff
@@ -67,12 +70,14 @@ static <T> T builder(Protocol<T> protocol) {
 
         private URI baseUri;
         private Tls tls;
-        private SocketOptions channelOptions;
+        private final SocketOptions.Builder socketOptionsBuilder = SocketOptions.builder();
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:73]: The existing field is called `channelOptions`, let's call the new one `channelOptionsBuilder`.

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 4]
```diff
@@ -183,6 +192,28 @@ public B maxRedirects(int maxRedirect) {
...
         }
 
+        /**
+         * Sets new connection timeout.
+         *
+         * @param timeout amount of time
+         * @return updated builder
+         */
+        public B connectTimeout(Duration timeout) {
+            this.connectTimeout = timeout;
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:202]: `channelOptionsBuilder.connectTimeout(timeout)`

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 5]
```diff
@@ -183,6 +192,28 @@ public B maxRedirects(int maxRedirect) {
...
+        }
+
+        /**
+         * Sets new read timeout.
+         *
+         * @param timeout amount of time
+         * @return updated builder
+         */
+        public B readTimeout(Duration timeout) {
+            this.readTimeout = timeout;
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:213]: `channelOptionsBuilder.readTimeout(timeout)`

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/ClientConfig.java [hunk 6]
```diff
@@ -44,6 +45,20 @@ public interface ClientConfig {
      */
     SocketOptions socketOptions();
 
+    /**
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/ClientConfig.java:48]: These two methods should not be needed.

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java (hunk 1 above)
Comment (aserkes) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:137]: Does it mean that I need to add to JavaDocs to the `readTimeout` and `connectTimeout` shortcuts the notes that they have less priority than `readTimeout` and `connectTimeout` from `channelOptions` and they will be discarded if `channelOptions` will be used during the build of the `WebClient`?

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java [hunk 7]
```diff
@@ -236,8 +281,11 @@ protected B removeHeader(Http.HeaderName name) {
          *
          * @return socket options
          */
-        SocketOptions channelOptions() {
-            return channelOptions;
+        protected SocketOptions channelOptions() {
+            if (channelOptions != null) {
```
Comment (romain-grecourt) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:285]: Can you do this as part of the build method instead ?

Comment (aserkes) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:285]: `WebClient.Builder` does not implement build method. 
Do I need to implement it? At the moment I added to the build method of the class `io.helidon.nima.webclient.http1.Http1Client.Http1ClientBuilder` that implements this interface the line : 
> configBuilder.socketOptions(super.channelOptions());

Do I need to replace this line by the code from the commented snippet above?

Comment (romain-grecourt) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/WebClient.java:298]: You can add something like this to `WebClient.Builder`:

```java
/**
 * Actual {@link #build()} implementation for {@link WebClient} subclasses.
 *
 * @return new client
 */
protected abstract C doBuild();

@Override
public C build() {
    // do some pre-build stuff here...
    return doBuild();
}
```

Then update the current implementations of `build` to be of `doBuild`.

**Review by romain-grecourt — APPROVED** [review]

**Pull request merged**